src/main/java/com/mixpanel/android/mpmetrics/Tweaks.java

```diff
@@ -33,18 +34,88 @@ public synchronized void addOnTweakDeclaredListener(OnTweakDeclaredListener list
     }
 
     /**
+     * Registers a {@link OnTweakUpdatedListener} to get notified about tweak changes
+     * @param listener the {@link OnTweakUpdatedListener} to be notified
+     */
+    public synchronized void addOnTweakUpdatedListener(OnTweakUpdatedListener listener) {
+        if (null == listener) {
+            throw new NullPointerException("listener cannot be null");
+        }
+        mTweakUpdatedListeners.add(listener);
+    }
+
+    /**
+     * Unregisters a {@link OnTweakUpdatedListener}
+     * @param listener the {@link OnTweakUpdatedListener} to be unregistered
+     */
+    public synchronized void removeOnTweakUpdatedListener(OnTweakUpdatedListener listener) {
+        mTweakUpdatedListeners.remove(listener);
+    }
+
+    /**
      * Manually set the value of a tweak. Most users of the library will not need to call this
      * directly - instead, the library will call set when new values of the tweak are published.
      */
     public synchronized void set(String tweakName, Object value) {
+        boolean autoDeclared = false;
+
         if (!mTweakValues.containsKey(tweakName)) {
-            Log.w(LOGTAG, "Attempt to set a tweak \"" + tweakName + "\" which has never been defined.");
-            return;
+            Log.i(LOGTAG, "Auto-declaring unknown tweak with name \"" + tweakName + "\"");
+            this.declareTweak(tweakName, value, this.detectTweakType(value));
+            autoDeclared = true;
+
         }
 
+
         final TweakValue container = mTweakValues.get(tweakName);
+        Object oldValue = container.value;
         final TweakValue updated = container.updateValue(value);
+        Object newValue = updated.value;
         mTweakValues.put(tweakName, updated);
+
+        if(autoDeclared || !oldValue.equals(newValue)) {
+
+            Log.i(LOGTAG, "Updated tweak with name \"" + tweakName + "\": " + updated.getStringValue());
+
+            for (OnTweakUpdatedListener listener : mTweakUpdatedListeners) {
```

Comment:
Why do you want to call the listener here? If you have multiple `Tweak`s, the listener will get called when you update the first `Tweak`, while all the other `Tweak`s are having the default value.